File: packages/core/src/diff-expand.ts
export const MAX_DIFF_EXPAND_BYTES = 2 * 1024 * 1024

export type DiffSite = {
  readonly oldText: string
  readonly newText: string
  readonly oldLine?: number
  readonly newLine?: number
}

export type DiffExpandInput = {
  readonly diskText: string | undefined
  readonly oldRegion: string
  readonly newRegion: string
  readonly diskIsBefore: boolean
  readonly replaceAll?: boolean
  readonly sites?: ReadonlyArray<DiffSite>
}

export type DiffSides = {
  readonly oldText: string
  readonly newText: string
  readonly firstChangedLine: number
  readonly wholeFile: boolean
}

import { utf8ByteLength } from "./utf8.js"

const utf8Bytes = utf8ByteLength

const oversize = (text: string | undefined): boolean =>
  text !== undefined && utf8Bytes(text) > MAX_DIFF_EXPAND_BYTES

const toLf = (text: string): string => text.replace(/\r\n/g, "\n")

const restoreCrlf = (lfText: string): string => lfText.replace(/\n/g, "\r\n")

const firstChangedLine = (oldText: string, newText: string): number => {
  const oldLines = oldText.split("\n")
  const newLines = newText.split("\n")
  const n = Math.min(oldLines.length, newLines.length)
  for (let i = 0; i < n; i++) {
    if (oldLines[i] !== newLines[i]) return i
  }
  return n === 0 ? 0 : n
}

const regionOnly = (oldRegion: string, newRegion: string): DiffSides => ({
  oldText: oldRegion,
  newText: newRegion,
  firstChangedLine: firstChangedLine(oldRegion, newRegion),
  wholeFile: false
})

const wholeFile = (oldText: string, newText: string): DiffSides => ({
  oldText,
  newText,
  firstChangedLine: firstChangedLine(oldText, newText),
  wholeFile: true
})

const indexesOf = (haystack: string, needle: string): Array<number> => {
  if (needle === "") return []
  const out: Array<number> = []
  let from = 0
  while (from <= haystack.length) {
    const at = haystack.indexOf(needle, from)
    if (at === -1) break
    out.push(at)
    from = at + Math.max(needle.length, 1)
  }
  return out
}

const lineAtOffset = (text: string, offset: number): number => {
  let line = 1
  const limit = Math.min(offset, text.length)
  for (let i = 0; i < limit; i++) {
    if (text[i] === "\n") line++
  }
  return line
}

const pickIndex = (
  haystack: string,
  needle: string,
  siteLine: number | undefined
): number => {
  const hits = indexesOf(haystack, needle)
  if (hits.length === 0) return -1
  if (siteLine === undefined) return hits[0] ?? -1
  const match = hits.find((at) => lineAtOffset(haystack, at) === siteLine)
  return match ?? hits[0] ?? -1
}

const replaceOnceAt = (haystack: string, start: number, needle: string, replacement: string): string =>
  haystack.slice(0, start) + replacement + haystack.slice(start + needle.length)

const replaceAll = (haystack: string, needle: string, replacement: string): string => {
  if (needle === "") return haystack
  return haystack.split(needle).join(replacement)
}

type Located = {
  readonly text: string
  readonly crlf: boolean
}

const locateHaystack = (haystack: string, needle: string): Located | undefined => {
  if (needle === "" || haystack.includes(needle)) return { text: haystack, crlf: haystack.includes("\r\n") }
  if (haystack.includes("\r\n") && !needle.includes("\r\n")) {
    const lf = toLf(haystack)
    if (lf.includes(needle)) return { text: lf, crlf: true }
  }
  return undefined
}

const emit = (located: Located): string => located.crlf && !located.text.includes("\r\n")
  ? restoreCrlf(located.text)
  : located.text

export const expandDiffToWholeFile = (input: DiffExpandInput): DiffSides => {
  const { diskText, oldRegion, newRegion, diskIsBefore } = input
  if (oversize(diskText) || oversize(oldRegion) || oversize(newRegion)) {
    return regionOnly(oldRegion, newRegion)
  }

  if (diskText === undefined) {
    if (oldRegion === "") return wholeFile("", newRegion)
    return regionOnly(oldRegion, newRegion)
  }

  if (oldRegion === "" && diskIsBefore && diskText.length > 0) {
    return wholeFile(diskText, newRegion)
  }

  const needle = diskIsBefore ? oldRegion : newRegion
  const replacement = diskIsBefore ? newRegion : oldRegion
  if (needle === "") {
    return diskIsBefore ? wholeFile(diskText, newRegion) : regionOnly(oldRegion, newRegion)
  }

  const located = locateHaystack(diskText, needle)
  if (located === undefined) return regionOnly(oldRegion, newRegion)

  const siteLine = input.sites?.[0]?.oldLine
  let nextText: string
  if (input.replaceAll === true) {
    nextText = replaceAll(located.text, needle, replacement)
  } else {
    const at = pickIndex(located.text, needle, siteLine)
    if (at === -1) return regionOnly(oldRegion, newRegion)
    nextText = replaceOnceAt(located.text, at, needle, replacement)
  }

  const proposed = emit({ text: nextText, crlf: located.crlf })
  const original = emit(located)
  return diskIsBefore ? wholeFile(original, proposed) : wholeFile(proposed, original)
}

File: beard/core/src/main/scala/groksbeard/core/DiffExpand.scala
package groksbeard.core

final case class DiffSite(
    oldText: String,
    newText: String,
    oldLine: Option[Int] = None,
    newLine: Option[Int] = None,
)

final case class DiffExpandInput(
    diskText: Option[String],
    oldRegion: String,
    newRegion: String,
    diskIsBefore: Boolean,
    replaceAll: Boolean = false,
    sites: List[DiffSite] = Nil,
)

final case class DiffSides(
    oldText: String,
    newText: String,
    firstChangedLine: Int,
    wholeFile: Boolean,
)

object DiffExpand:
  val MaxBytes: Int = 2 * 1024 * 1024

  def expand(input: DiffExpandInput): DiffSides =
    val DiffExpandInput(diskText, oldRegion, newRegion, diskIsBefore, replaceAll, sites) = input
    if oversize(diskText) || oversize(Some(oldRegion)) || oversize(Some(newRegion)) then
      regionOnly(oldRegion, newRegion)
    else
      diskText match
        case None =>
          if oldRegion.isEmpty then wholeFile("", newRegion) else regionOnly(oldRegion, newRegion)
        case Some(disk) =>
          if oldRegion.isEmpty && diskIsBefore && disk.nonEmpty then wholeFile(disk, newRegion)
          else
            val needle      = if diskIsBefore then oldRegion else newRegion
            val replacement = if diskIsBefore then newRegion else oldRegion
            if needle.isEmpty then if diskIsBefore then wholeFile(disk, newRegion) else regionOnly(oldRegion, newRegion)
            else
              locateHaystack(disk, needle) match
                case None          => regionOnly(oldRegion, newRegion)
                case Some(located) =>
                  val siteLine = sites.headOption.flatMap(_.oldLine)
                  if replaceAll then
                    val nextText = replaceEvery(located.text, needle, replacement)
                    sidesFrom(located, nextText, diskIsBefore)
                  else
                    val at = pickIndex(located.text, needle, siteLine)
                    if at < 0 then regionOnly(oldRegion, newRegion)
                    else
                      val nextText = replaceOnceAt(located.text, at, needle, replacement)
                      sidesFrom(located, nextText, diskIsBefore)
            end if
    end if
  end expand

  private def oversize(text: Option[String]): Boolean =
    text.exists(t => Utf8.byteLength(t) > MaxBytes)

  private def firstChangedLine(oldText: String, newText: String): Int =
    val oldLines = oldText.split("\n", -1)
    val newLines = newText.split("\n", -1)
    val n        = math.min(oldLines.length, newLines.length)
    var i        = 0
    while i < n do
      if oldLines(i) != newLines(i) then return i
      i += 1
    if n == 0 then 0 else n

  private def regionOnly(oldRegion: String, newRegion: String): DiffSides =
    DiffSides(oldRegion, newRegion, firstChangedLine(oldRegion, newRegion), wholeFile = false)

  private def wholeFile(oldText: String, newText: String): DiffSides =
    DiffSides(oldText, newText, firstChangedLine(oldText, newText), wholeFile = true)

  private def sidesFrom(located: Located, nextText: String, diskIsBefore: Boolean): DiffSides =
    val proposed = emit(located.copy(text = nextText))
    val original = emit(located)
    if diskIsBefore then wholeFile(original, proposed) else wholeFile(proposed, original)

  private def indexesOf(haystack: String, needle: String): List[Int] =
    if needle.isEmpty then Nil
    else
      val out  = List.newBuilder[Int]
      var from = 0
      while from <= haystack.length do
        val at = haystack.indexOf(needle, from)
        if at < 0 then from = haystack.length + 1
        else
          out += at
          from = at + math.max(needle.length, 1)
      out.result()

  private def lineAtOffset(text: String, offset: Int): Int =
    var line  = 1
    val limit = math.min(offset, text.length)
    var i     = 0
    while i < limit do
      if text.charAt(i) == '\n' then line += 1
      i += 1
    line

  private def pickIndex(haystack: String, needle: String, siteLine: Option[Int]): Int =
    val hits = indexesOf(haystack, needle)
    if hits.isEmpty then -1
    else
      siteLine match
        case None       => hits.head
        case Some(line) =>
          hits.find(at => lineAtOffset(haystack, at) == line).getOrElse(hits.head)

  private def replaceOnceAt(haystack: String, start: Int, needle: String, replacement: String): String =
    haystack.substring(0, start) + replacement + haystack.substring(start + needle.length)

  private def replaceEvery(haystack: String, needle: String, replacement: String): String =
    if needle.isEmpty then haystack else haystack.replace(needle, replacement)

  private final case class Located(text: String, crlf: Boolean)

  private def locateHaystack(haystack: String, needle: String): Option[Located] =
    if needle.isEmpty || haystack.contains(needle) then Some(Located(haystack, haystack.contains("\r\n")))
    else if haystack.contains("\r\n") && !needle.contains("\r\n") then
      val lf = haystack.replace("\r\n", "\n")
      if lf.contains(needle) then Some(Located(lf, crlf = true)) else None
    else None

  private def emit(located: Located): String =
    if located.crlf && !located.text.contains("\r\n") then located.text.replace("\n", "\r\n")
    else located.text
end DiffExpand
